lib/rust/profiler/src/lib.rs

```diff
@@ -0,0 +1,333 @@
+        // It takes (2 ** 64) * 100us = 58_455_453 years for a Timestamp to wrap.
+        unsafe { Self(num::NonZeroU64::new_unchecked(ticks + 1)) }
+    }
+
+    /// Convert to an offset from the time origin, in ms.
+    pub fn into_ms(self) -> f64 {
+        (self.0.get() - 1) as f64 / 10.0
+    }
+}
+
+// === FFI ===
+
+pub mod js {
+    pub mod performance {
+        use wasm_bindgen::prelude::*;
+
+        #[wasm_bindgen]
+        extern "C" {
+            #[wasm_bindgen(js_namespace = performance)]
+            pub fn now() -> f64;
+        }
+    }
+}
+
+
+
+// ==================
+// === ProfilerId ===
+// ==================
+
+/// Uniquely identifies a runtime measurement.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct ProfilerId(u32);
+
+impl ProfilerId {
+    fn new() -> Self {
+        thread_local! {
+            pub static NEXT_ID: cell::Cell<u32> = cell::Cell::new(1);
+        }
+        ProfilerId(NEXT_ID.with(|next_id| {
+            let id = next_id.get();
+            next_id.set(id + 1);
+            id
+        }))
+    }
+}
+
+/// Pseudo-profiler serving as the root of the measurement hierarchy.
+pub const APP_LIFETIME: Objective = Objective(ProfilerId(0));
+
+
+
+// =======================
+// === StartedProfiler ===
+// =======================
+
+/// The interface supported by profiler-data objects.
+pub trait StartedProfiler {
+    fn finish(&self, profiler: ProfilerId);
+}
+
+// === Implementation for disabled profilers ===
+
+impl StartedProfiler for () {
+    fn finish(&self, _: ProfilerId) {}
+}
+
+// ====================
+// === ProfilerData ===
+// ====================
+
+/// Data used by a started Measurement for an enabled profile level.
+pub struct ProfilerData {
+    pub parent: ProfilerId,
+    pub start:  Option<Timestamp>,
+    pub label:  Label,
+}
+
+// === Trait Implementations ===
+
+impl StartedProfiler for ProfilerData {
+    fn finish(&self, profiler: ProfilerId) {
+        let parent = self.parent;
+        let start = self.start;
+        let label = self.label;
+        let end = Timestamp::now();
+        let measurement = Measurement { parent, profiler, start, end, label };
+        crate::MEASUREMENTS.with(move |log| log.push(measurement));
+    }
+}
+
+
+
+// ===================
+// === Measurement ===
+// ===================
+
+/// Identifies a profiled section, the parent it was reached by, and its entry and exit times.
+pub struct Measurement {
+    pub parent:   ProfilerId,
+    pub profiler: ProfilerId,
+    pub start:    Option<Timestamp>,
+    pub end:      Timestamp,
+    pub label:    Label,
+}
+
+pub mod internal {
+    use crate::*;
+
+    use std::cell;
+    use std::mem;
+
+    // =======================
+    // === LocalVecBuilder ===
+    // =======================
+
+    /// Data structure supporting limited interior mutability, to build up a collection.
+    pub struct LocalVecBuilder<T>(cell::UnsafeCell<Vec<T>>);
```

Comment:
This is a cool structure! Maybe it should be extracted to `enso/lib/rust/data-structures`?

Reply:
I'll do that. Another reusable component of this: `wrap_body` could easily be used to implement a macro that automatically applies the inner-monomorphic-function pattern. It could be used like this:

```rust
// The attribute accepts the names of common traits that fit a many-input-types,
// one-output-type pattern (mainly Into and AsRef); it effectively moves the polymorphic
// conversion to the call site, so that the body doesn't need to be generic over the given traits.
#[monomorphic(Into)]
fn foo(x: impl Into<u32>, y: impl Into<u32>) {
    // function body will only be compiled once,
    // instead of for every combination of (typeof x) * (typeof y)
}
```

Because it's easy to implement, easy to apply, and helps with monomorphization (which is one of the biggest causes of code bloat), this is probably one of the lowest-hanging fruit of compile-size optimizations. For now, I'll just put `wrap_body` in its own crate so that we'll have it handy when we get to the optimizing phase.